climada/util/coordinates.py

```diff
@@ -1554,21 +1569,32 @@ def read_raster(file_name, band=None, src_crs=None, window=None, geometry=None,
 
     return dst_meta, intensity.reshape(dst_shape)
 
-def read_raster_bounds(path, bounds, res=None, bands=None):
+def read_raster_bounds(path, bounds, res=None, bands=None, resampling="nearest",
```

Comment:
It's not two actions, but it's a single call to `rasterio.warp.reproject`.

Reply:
Fair point.

Reply:
Is there much difference in the logic between `read_raster` and `read_raster_bounds`? From a skim, the latter is more computationally efficient, but they do similar things (though one is read-reproject-crop and the other is read-crop-reproject).

I do wonder if there would be benefit (in the future) in atomising these methods to simpler operations that you can chain together. e.g.:
```
ras = u_coord.Raster.from_file(path)
  .reproject(...)
  .crop(...)
```

Reply:
The function `read_raster_bounds` does neither `read-reproject-crop` nor `read-crop-reproject`. It's all in one call to the `rasterio.warp.reproject` function. Furthermore, it crops using specified bounds in terms of CRS coordinates and gives control about the added cell padding.

The function `read_raster` did already exist when I started as a contributor of CLIMADA. Back then, it was already much to complex for a typical helper function. That's why I moved one part of its source code to a non-API function `_read_raster_reproject`. But still, it's very hard to understand what this function really does when several kwargs (like window, dst_crs, transform, width/height and geometry) are specified by the user. As you said, the `read_raster` function already has some crop capabilities: It can crop according to a given `rasterio.window.Window` (which is in terms of the source raster's grid cells), it can crop and mask according to a given `shapely.geometry`, or it can crop according to a given `transform`, `width`, `height` specification.

When I added the `read_raster_bounds` function, I intentionally decided to add a new function instead of adding new functionality to the already overly complex `read_raster` function. That function already has too many concurrent kwargs, and way too many lines of code.

I'm clearly in favor of simplifying the `read_raster_*` functions in CLIMADA. However, that's no easy task, given the numerous applications out there. And it's clearly out of scope for this PR.

Reply:
Thanks for the clear answer!

I definitely misunderstood `read_raster_bounds` when I read it through - you've cleared it up for me now, thanks.